pkg/ruler/api_test.go

```diff
@@ -337,24 +337,74 @@ rules:
 `,
 			output: "per-user rules per rule group limit (limit: 1 actual: 2) exceeded\n",
 		},
+	}
+
+	for _, tt := range tc {
+		t.Run(tt.name, func(t *testing.T) {
+			router := mux.NewRouter()
+			router.Path("/api/v1/rules/{namespace}").Methods("POST").HandlerFunc(a.CreateRuleGroup)
+			// POST
+			req := requestFor(t, http.MethodPost, "https://localhost:8080/api/v1/rules/namespace", strings.NewReader(tt.input), "user1")
+			w := httptest.NewRecorder()
+
+			router.ServeHTTP(w, req)
+			require.Equal(t, tt.status, w.Code)
+			require.Equal(t, tt.output, w.Body.String())
+		})
+	}
+}
+
+func TestRuler_RulerGroupLimits(t *testing.T) {
+	cfg, cleanup := defaultRulerConfig(newMockRuleStore(make(map[string]rules.RuleGroupList)))
+	defer cleanup()
+
+	r, rcleanup := newTestRuler(t, cfg)
+	defer rcleanup()
+	defer services.StopAndAwaitTerminated(context.Background(), r) //nolint:errcheck
+
+	r.limits = &ruleLimits{maxRuleGroups: 1, maxRulesPerRuleGroup: 1}
+
+	a := NewAPI(r, r.store)
+
+	tc := []struct {
+		name   string
+		input  string
+		output string
+		err    error
+		status int
+	}{
+		{
+			name:   "when pushing the first group within bounds of the limit",
+			status: 202,
+			input: `
+name: test_first_group_will_succeed
+interval: 15s
+rules:
+- record: up_rule
+  expr: up{}
+`,
+			output: "{\"status\":\"success\",\"data\":null,\"errorType\":\"\",\"error\":\"\"}",
+		},
 		{
-			name:   "when exceeding the rule group limit",
+			name:   "when exceeding the rule group limit after sending the first group",
 			status: 400,
 			input: `
-name: test
+name: test_second_group_will_fail
 interval: 15s
 rules:
 - record: up_rule
   expr: up{}
 `,
-			output: "per-user rules per rule group limit (limit: 1 actual: 1) exceeded\n",
+			output: "per-user rule groups limit (limit: 1 actual: 2) exceeded\n",
 		},
 	}
 
 	for _, tt := range tc {
+		// define once so the requests build on each other so the number of rules can be tested
+		router := mux.NewRouter()
+		router.Path("/api/v1/rules/{namespace}").Methods("POST").HandlerFunc(a.CreateRuleGroup)
+
```

Comment:
I don't understand this change.  It's not run "once" - it's done again for each element in `tc`.
And I can't see what practical difference moving it outside of `Run()` makes.

Reply:
Right you are, definitely not going to work that way :)